Propagate errors from DSW to pod events

jsafrane opened the pull request:
**What this PR does / why we need it**:
Goal: tell users that pod cannot run because:
 - PVC does not exist
 - PVC is not bound
 - PVC uses a different volumeMode than the pod expects.
 - ...

The first two bullets are checked also by scheduler, but kubelet should report errors here too, in case a pod runs with a different scheduler or was scheduled manually.

I added list of errors for each pod to DesiredStateOfWorld (DSW), where DSW populator (DSWP) stores them and kubelet picks them up via `WaitForAttachAndMount` and creates pod event(s) from them.

I did not want to add errors to `dsw.volumesToMount`, because that would create `volumesToMount` entries with incomplete / erroneous volumes (e.g. PVC was not bound yet and we don't know unique volume name).

I hope I got the cleaning right - DSWP periodically checks all stored errors and removes those for deleted pods.

I cleaned some error messages on the way, so the events are nice and tidy (well, at least they're better than before, see below).

Natural extension would be to report also mount / set up errors from reconciler ((not part of this PR). By placing the list of errors to DSW, there is no way to pass errors from unmount / teardown (there is no entry in DSW for pods that are being torn down).

**Examples** (simulating narrow terminal window):

* PVC does not exist:

  ```
  Type     Reason       Age   From                Message
  ----     ------       ----  ----                -------
  Warning  FailedMount  9s    kubelet, 127.0.0.1  Unable to attach or mount volumes:
  error processing PVC default/myclaim: failed to fetch PVC from API server:
  persistentvolumeclaims "myclaim" not found. List of unmounted volumes=[vol]. List 
  of unattached volumes=[vol default-token-4pt45].
  ```

* PVC is not bound:
  ```
  Type     Reason       Age   From                Message
  ----     ------       ----  ----                -------
  Warning  FailedMount  4s    kubelet, 127.0.0.1  Unable to attach or mount volumes:
  error processing PVC default/myclaim: PVC is not bound. List of unmounted volumes=
  [vol]. List of unattached volumes=[vol default-token-4pt45]. 
  ```

* Wrong mode:
  ```
  Type     Reason                  Age   From                     Message
  ----     ------                  ----  ----                     -------
  Normal   SuccessfulAttachVolume  4s    attachdetach-controller  AttachVolume.Attach 
  succeeded for volume "mypv"
  Warning  FailedMount             4s    kubelet, 127.0.0.1       Unable to attach or
  mount volumes: volume vol has volumeMode Filesystem, but is specified in 
  volumeDevices. List of unmounted volumes=[vol default-token-4pt45]. List of 
  unattached volumes=[vol default-token-4pt45].
  ```

* Multiple errors:
  ```
  Type     Reason       Age   From                Message
  ----     ------       ----  ----                -------
  Warning  FailedMount  1s    kubelet, 127.0.0.1  Unable to attach or mount volumes: 
  error processing PVC default/myclaim2: failed to fetch PVC from API server: 
  persistentvolumeclaims "myclaim2" not found; error processing PVC default/myclaim: 
  failed to fetch PVC from API server: persistentvolumeclaims "myclaim" not found. 
  List of unmounted volumes=[vol vol2 default-token-5jpm2]. List of unattached 
  volumes=[vol vol2 default-token-5jpm2].
  ```

I might send every error as a separate event. Would it be too many events? Does 
`List of unmounted volumes=[vol vol2 default-token-5jpm2]. List of unattached 
  volumes=[vol vol2 default-token-5jpm2].` have any value?


**Which issue(s) this PR fixes**:
Fixes #79794
 
**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".
-->
```release-note
Errors from pod volume set up are now propagated as pod events.
```

/kind bug
/sig storage

@msau42 @jingxu97 @gnufied, PTAL

@mkimuram, I am afraid that checking for events in #79796 is the best we can do here.

jsafrane committed:
Add events to dswp

mkimuram commented on pkg/kubelet/volumemanager/volume_manager.go:
```diff
@@ -375,9 +376,10 @@ func (vm *volumeManager) WaitForAttachAndMount(pod *v1.Pod) error {
+			"failed to attach or mount for pod %q/%q: %s. List of unmounted volumes=%v, list of unattached volumes=%v.",
 			pod.Namespace,
 			pod.Name,
+			err,
```
For readability in multiple errors case, would it be better to output errors after lists of volumes?
(It might be too much to add "List of errors=" before errors, though.)

jsafrane replied:
Ack, I reordered the messages as you suggest:

```
Events:
  Type     Reason       Age   From                Message
  ----     ------       ----  ----                -------
  Warning  FailedMount  1s    kubelet, 127.0.0.1  Unable to attach or mount 
volumes: unmounted volumes=[vol], unattached volumes=[vol default-token-p67zc]: 
error processing PVC default/myclaim: failed to fetch PVC from API server: 
persistentvolumeclaims "myclaim" not found
```

msau42 commented:
/assign @jingxu97

gnufied commented on pkg/kubelet/volumemanager/volume_manager.go:
```diff
@@ -402,6 +402,9 @@ func (vm *volumeManager) getUnattachedVolumes(expectedVolumes []string) []string
 func (vm *volumeManager) verifyVolumesMountedFunc(podName types.UniquePodName, expectedVolumes []string) wait.ConditionFunc {
 	return func() (done bool, err error) {
+		if errs := vm.desiredStateOfWorld.PopPodErrors(podName); len(errs) > 0 {
+			return true, errors.New(strings.Join(errs, "; "))
```
should it return false from here?

jsafrane replied:
It does not really matter, `wait.Poll` will report error in both cases.

gnufied commented on pkg/kubelet/volumemanager/populator/desired_state_of_world_populator.go:
```diff
@@ -309,11 +318,12 @@ func (dswp *desiredStateOfWorldPopulator) processPodVolumes(
 				volumeSpec.Name(),
 				uniquePodName,
 				err)
+			dswp.desiredStateOfWorld.AddErrorToPod(uniquePodName, err.Error())
```
If for some reason if these errors were transient and volume eventually did get added to DSOW.. should we remove the errors from pod if `allVolumesAdded` is true in lin#343?

jsafrane replied:
Good idea, I remove errors if `allVolumesAdded`.

jsafrane committed:
Fix some error messages

jsafrane commented:
/retest

gnufied commented:
Looks good to me. but I want @jingxu97 to have a look as well.

/lgtm
/hold

gnufied commented:
/priority important-longterm

gnufied commented:
/hold cancel

Assigning to @derekwaynecarr for approval

/assign @derekwaynecarr

jingxu97 commented on pkg/kubelet/volumemanager/volume_manager_test.go:
```diff
@@ -124,9 +125,16 @@ func TestInitialPendingVolumesForPodAndGetVolumesInUse(t *testing.T) {
 	go delayClaimBecomesBound(kubeClient, claim.GetNamespace(), claim.ObjectMeta.Name)
 
-	err = manager.WaitForAttachAndMount(pod)
+	err = wait.Poll(100*time.Millisecond, 1*time.Second, func() (bool, error) {
```
without this wait.Poll, the test should be flaky?

jsafrane replied:
Yes, because the PVC is bound after a short delay (see above) and the first few calls return error that the PVC is not bound yet.

jingxu97 commented:
just one thing since the error might be transient, for example, PVC is not created whey pod is created. But then PVC is added to desired state, but fail to attach/mount for other reason. When user see the events, will it cause some confusion on what is the reason that pod fails to start?

jsafrane commented:
I thinks that regular volume operation errors are reported as events too, so user will see say "PVC does not exist" and "SetUp failed: volume not found" events and should know where is the pod stuck at.

jingxu97 commented:
/lgtm

jsafrane commented:
/assign @derekwaynecarr @tallclair 
for approval on sig-node side (`verifyVolumesMountedFunc` changes a bit).

derekwaynecarr commented on pkg/kubelet/volumemanager/cache/desired_state_of_world.go:
```diff
@@ -412,3 +430,36 @@ func (dsw *desiredStateOfWorld) isDeviceMountableVolume(volumeSpec *volume.Spec)
 	return false
 }
+
+func (dsw *desiredStateOfWorld) AddErrorToPod(podName types.UniquePodName, err string) {
```
how big can this get?  is it possible for this to grow unbounded ?

jsafrane replied:
Good question, errors are stored in `sets.String`, so it *can* grow if every error has a different text. I added a limit of 10, just in case there is a bug in the populator.

derekwaynecarr reviewed (commented):
have a question, once clarified, this is good from my side.

jsafrane committed:
Add limit of stored errors

There should be a limit so the set of errors does not grow too much in
case of some bug in the populator.

jsafrane commented:
/retest

derekwaynecarr commented:
thanks for bounding the set.

/approve
/lgtm
